[AMDGPU] Fix canonicalization of truncated values.

hvdijk opened the pull request:
We were relying on roundings to implicitly canonicalize, which is generally safe, except with roundings that may be optimized away.

Fixes #82937.

hvdijk commented:
My first attempt for this, special casing `FP_ROUND` operands, worked for the tests we had, but not in the general case, because it did not take into account that we may have nodes that will later be optimized to just `FP_ROUND`. What this second attempt does is, if we determine that an operand is already canonicalized, keep track of whether we saw any `FP_ROUND` with `TRUNC=1`. If we did, then we can still optimize `FCANONICALIZE` to a no-op by inserting a `FREEZE` node to block optimizations. This should be correct in all cases and avoids redundant recanonicalizations in the most common cases, but as the test updates show, does not avoid redundant recanonicalizations in all cases. I will attempt to update this further.

arsenm commented on llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:
```diff
@@ -26258,6 +26258,24 @@ SDValue DAGCombiner::visitFP_TO_FP16(SDNode *N) {
   if (N0->getOpcode() == ISD::FP16_TO_FP)
     return N0->getOperand(0);
 
+  // fold (fp_to_fp16 (freeze (fp16_to_fp (fp_to_fp16 op))))
```
(Commenting in the right place this time)

This is a specific instance of isGuaranteedNotToBePoison; it would be better to reframe this in terms of a new DAG clone of the ValueTracking helper

jayfoad replied:
> I'm slightly abusing the lack of freeze optimisations to use freeze as a general optimisation barrier.

That does not sound safe to me.

hvdijk replied:
> That does not sound safe to me.

Understandable. It's still safer than the status quo where we miscompile code and have tests to ensure that we miscompile code, but if we can do better that would be great.

The two options I see are adding that dummy node type and then treating that exactly as I am currently treating `FREEZE`, or just saying that `FP_ROUND` and `FP_EXTEND` do not canonicalise and ignoring any performance degradations due to redundant recanonicalisations. Would either of those approaches be acceptable? Or is there some other option that I am just not seeing?

jayfoad replied:
> > That does not sound safe to me.
> 
> Understandable.

That was a bit of a drive-by comment (sorry) since I am not deeply involved in this code. I'm not sure I understand the underlying problem. Why is it OK to optimize away a rounding, if we are relying on it also canonicalizing? Is this because we're running in a mode where general fp operations are not required to canonicalize, but we do want to preserve explicit canonicalization ops?

hvdijk replied:
Yes, exactly that. Aside from SNaN handling, `(fp_round (fp_extend x))` and `(fp_extend (fp_round x, /*trunc*/1))` are always equal to `x` and so for code that does not care about SNaNs, these are useful to optimise away. But `(fp_round (fcanonicalize (fp_extend x)))` and `(fp_extend (fcanonicalize (fp_round x, /*trunc*/1)))` must handle SNaNs, so we cannot just optimise the `fcanonicalize` to a no-op on the basis that `fp_extend`/`fp_round` will already have canonicalized.

jayfoad replied:
Off the top of my head we could either:
1. Add a "must canonicalize" version of fp_round and/or fp_extend. But this does not sound very extendable, since we might also need "must canonicalize" versions of any other fp op that could be optimized away. Or:
2. Do not optimize away fcanonicalize as a DAG combine, but detect at instruction selection time when it is redundant because the input is already canonicalized. I guess this is what @arsenm suggested earlier (sorry I'm just catching up now). Is it hard to do this (without significant regressions)?

hvdijk replied:
1) indeed does not sound very extendable, and I suspect we will get more optimisable operations in the future.
2) is what @arsenm briefly suggested earlier, before suggesting that maybe it should be done even later. It's very hard to do without regressions, but whether any of the regressions will count as *significant* is hard to say in advance.

hvdijk replied:
Let's just bite the bullet and do it, I guess. Some of the regressions are easily solved and I have included them in the same commit. Others are not but the degradations that remain actually seem like they are outweighed by better `fcanonicalize` optimisations elsewhere.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -12416,7 +12416,7 @@ SDValue SITargetLowering::performRcpCombine(SDNode *N,
 
 bool SITargetLowering::isCanonicalized(SelectionDAG &DAG, SDValue Op,
-                                       unsigned MaxDepth) const {
+                                       bool &Trunc, unsigned MaxDepth) const {
```
If you always introduce a freeze whenever eliding any canonicalize, shouldn't need this special casing of trunc anymore?

hvdijk replied:
You're right, we can do that, it results in correct code, but that causes regressions where we are no longer able to figure out that we have an already canonicalised value and redundantly recanonicalise. I see that happen in the existing `clamp.ll`, `fmed3.ll`, `known-never-snan.ll`, `mad-mix-lo.ll`, and `minmax.ll` tests.

hvdijk replied:
With this different approach based on your idea, this is no longer needed.

jayfoad commented on llvm/lib/Target/AMDGPU/SIInstructions.td:
```diff
@@ -3590,6 +3618,17 @@ FPMinMaxPat<Instruction minmaxInst, ValueType vt, SDPatternOperator min_or_max,
+class
+FPMinCanonMaxPat<Instruction minmaxInst, ValueType vt, SDPatternOperator min_or_max,
+                 SDPatternOperator max_or_min_oneuse> : GCNPat <
+  (min_or_max (is_canonicalized_1<fcanonicalize>
```
I wonder how many more fp patterns there are that might need to skip over an `fcanonicalize` in the middle.

arsenm replied:
It's really arbitrary operations. fmin/fmax are just of particular interest because we're stuck inserting canonicalizes in the lowering for them. The more I think about it, the more I think we'd better of making this a separate transform (but just moving this to selection patterns is a net improvement for now I think)

hvdijk replied:
There may well be more, I just went for the ones that I came across from the test results. But if there are more, it doesn't result in wrong code, only suboptimal code, which I think is acceptable if the overall code still looks better?

arsenm replied:
I'd only consider the min/max legalize case important. There's only a handful of canonicalizes in the math library, they aren't common (and I think we could avoid most of those too if we rewrote the code slightly)

jayfoad reviewed (commented):
I like this approach. Makes a lot of sense to me.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -12574,6 +12595,9 @@ bool SITargetLowering::isCanonicalized(SelectionDAG &DAG, SDValue Op,
     return false;
 
   case ISD::BITCAST:
+    // TODO: This is incorrect as it loses track of the operand's type. We may
+    // end up effectively bitcasting from f32 to v2f16 or vice versa, and the
+    // same bits that are canonicalized in one type need not be in the other.
```
should give up on different element size bitcasts? (separate patch though)

hvdijk replied:
I think rather than give up, we should add a parameter specifying the type so that we would be able to handle things like `v2f16` -> `i32` -> truncate to `i16` -> `f16`. But yes, separate patch, this is a pre-existing issue.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -12603,20 +12627,21 @@ bool SITargetLowering::isCanonicalized(SelectionDAG &DAG, SDValue Op,
     case Intrinsic::amdgcn_trig_preop:
     case Intrinsic::amdgcn_log:
     case Intrinsic::amdgcn_exp2:
+    case Intrinsic::amdgcn_sqrt:
```
unrelated bonus change?

hvdijk replied:
Not unrelated, this is one of the things that came up in LLVM's testsuite where previously we were able to optimize away a `fcanonicalize` because we saw `sqrt`, but now we only check later and see `amdgcn_sqrt`.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstructions.td:
```diff
@@ -194,7 +194,26 @@ class HasOneUseTernaryOp<SDPatternOperator op> : PatFrag<
+    const SITargetLowering *TLI = static_cast<const SITargetLowering *>(
+      MF.getSubtarget().getTargetLowering());
+
+    return TLI->isCanonicalized(MI.getOperand(1).getReg(), const_cast<MachineFunction&>(MF));
```
why does isCanonicalized have a non-const MF parameter?

hvdijk replied:
Good question, I can take a look at that later today. This is basically just copied and pasted from the existing `is_canonicalized` form that already uses `const_cast`, but if we can change that, that would be good.

hvdijk replied:
It turns out there was no reason for that. Updated the PR to change it.

arsenm commented on llvm/lib/Target/AMDGPU/SIInstructions.td:
```diff
@@ -2944,6 +2944,34 @@ def : GCNPat<
+// If fcanonicalize's operand is implicitly canonicalized, we only need a copy.
+let AddedComplexity = 1000 in {
+def : GCNPat<
+  (is_canonicalized_1<fcanonicalize> f16:$src),
```
The PatFrag is what owns the predicate; it might be possible to define something using srcvalue as a way of getting a catch-all matcher with a predicate

hvdijk replied:
I had a look and couldn't find any custom predicate logic other than for instructions, which means we cannot do it on the operand level as the operand need not be an instruction. It's possible I missed something simple, but I think I would prefer to just leave it like this over making big changes to make TableGen support operand-level predicates.

hvdijk committed:
[AMDGPU] Fix canonicalization of truncated values.

We were relying on roundings to implicitly canonicalize, which is
generally safe, except that roundings may be optimized away, and more
generally, other operations may be optimized away as well.

To solve this, as suggested by @arsenm, keep fcanonicalize nodes around
for longer. Some tests revealed cases where we no longer figured out
that previous operations already produced canonicalized results but we
could easily change that; this commit includes those changes. Other
tests revealed cases where we no longer figure out that previous
operations already produced canonicalized results but larger changes are
needed to detect that; this commit disables those tests or updates the
expected results.

Fixes #82937.

arsenm reviewed (approved):
I think the tablegen can be simplified today, but it shouldn't block this patch

hvdijk commented:
Thanks! If I happen to end up figuring out how to simplify the TableGen I can create a followup PR, otherwise I look forward to seeing how someone else manages to do :)

arsenm commented:
> Thanks! If I happen to end up figuring out how to simplify the TableGen I can create a followup PR, otherwise I look forward to seeing how someone else manages to do :)

I managed to get it to work with this:

```
def is_canonicalized : PatLeaf<(fAny srcvalue:$src),
[{
    const SITargetLowering &Lowering =
              *static_cast<const SITargetLowering *>(getTargetLowering());
    return Lowering.isCanonicalized(*CurDAG, SDValue(N, 0));
}]
>;

def : GCNPat<
  (fcanonicalize (f32 is_canonicalized:$src)),
   (COPY f32:$src)
>;
```

jayfoad commented:
> > Thanks! If I happen to end up figuring out how to simplify the TableGen I can create a followup PR, otherwise I look forward to seeing how someone else manages to do :)
> 
> I managed to get it to work with this:
> 
> ```
> def is_canonicalized : PatLeaf<(fAny srcvalue:$src),
> [{
>     const SITargetLowering &Lowering =
>               *static_cast<const SITargetLowering *>(getTargetLowering());
>     return Lowering.isCanonicalized(*CurDAG, SDValue(N, 0));
> }]
> >;
> 
> def : GCNPat<
>   (fcanonicalize (f32 is_canonicalized:$src)),
>    (COPY f32:$src)
> >;
> ```

Awesome! I wonder if we could do something similar for is_divergent and is_uniform, so we don't need separate DivergentUnaryFrag/DivergentBinFrag/etc.

arsenm commented:
> Awesome! I wonder if we could do something similar for is_divergent and is_uniform, so we don't need separate DivergentUnaryFrag/DivergentBinFrag/etc.

Probably not. I'm running into a few issues trying to clean up the other uses. First, the predicate seems to not actually get applied when this is used as the source of a complex pattern. Plus we need at least another variant with an operand for the non-leaf cases
